src/transport/iroh_net.rs

```diff
@@ -0,0 +1,750 @@
+        Self::new_with_access_control(endpoint, AccessControl::Unrestricted)
+    }
+
+    /// Create a new server endpoint, with specified access control
+    ///
+    /// The server channel will take care of listening on the endpoint and spawning
+    /// handlers for new connections.
+    pub fn new_with_access_control(
+        endpoint: iroh_net::Endpoint,
+        access_control: AccessControl,
+    ) -> io::Result<Self> {
+        let allowed_node_ids = match access_control {
+            AccessControl::Unrestricted => BTreeSet::new(),
+            AccessControl::Allowed(list) if list.is_empty() => {
+                return Err(io::Error::other(
+                    "Empty list of allowed nodes, \
+                    endpoint would reject all connections",
+                ));
+            }
+            AccessControl::Allowed(list) => BTreeSet::from_iter(list),
+        };
+
+        let (ipv4_socket_addr, maybe_ipv6_socket_addr) = endpoint.bound_sockets();
+        let (sender, receiver) = flume::bounded(16);
+        let task = tokio::spawn(Self::endpoint_handler(
+            endpoint.clone(),
+            sender,
+            allowed_node_ids,
+        ));
+
+        Ok(Self {
+            inner: Arc::new(ListenerInner {
+                endpoint: Some(endpoint),
+                task: Some(task),
+                local_addr: once(LocalAddr::Socket(ipv4_socket_addr))
+                    .chain(maybe_ipv6_socket_addr.map(LocalAddr::Socket))
+                    .collect(),
+                receiver,
+            }),
+            _p: PhantomData,
+        })
+    }
+
+    /// Create a new server channel, given just a source of incoming connections
+    ///
+    /// This is useful if you want to manage the quinn endpoint yourself,
+    /// use multiple endpoints, or use an endpoint for multiple protocols.
+    pub fn handle_connections(
+        incoming: flume::Receiver<quinn::Connection>,
+        local_addr: SocketAddr,
+    ) -> Self {
+        let (sender, receiver) = flume::bounded(16);
+        let task = tokio::spawn(async move {
+            // just grab all connections and spawn a handler for each one
+            while let Ok(connection) = incoming.recv_async().await {
+                tokio::spawn(Self::connection_handler(connection, sender.clone()));
+            }
+        });
+        Self {
+            inner: Arc::new(ListenerInner {
+                endpoint: None,
+                task: Some(task),
+                local_addr: vec![LocalAddr::Socket(local_addr)],
+                receiver,
+            }),
+            _p: PhantomData,
+        }
+    }
+
+    /// Create a new server channel, given just a source of incoming substreams
+    ///
+    /// This is useful if you want to manage the quinn endpoint yourself,
+    /// use multiple endpoints, or use an endpoint for multiple protocols.
+    pub fn handle_substreams(
+        receiver: flume::Receiver<SocketInner>,
```

Comment:
I was confused since a public fn takes a SocketInner, but SocketInner is just a type alias for something public, right?

Reply:
Yeah exactly, `type SocketInner = (quinn::SendStream, quinn::RecvStream);` this part I also took from quinn transport. Although I don't it would be as useful as it is with quinn transport. I can remove both methods and leave them only on quinn transport.